OpenRA.Mods.Common/SpriteLoaders/GifLoader.cs

```diff
@@ -0,0 +1,315 @@
+
+				if (interlaced)
+				{
+					var originData = data;
+
+					var offsets = new int[]{0, 4, 2, 1};
+					var steps = new int[]{8, 8, 4, 2};
+
+					var j = 0;
+					for (var pass = 0; pass < 4; pass++)
+					{
+						for (var i = offsets[pass]; i < originData.Length; i += steps[pass])
+						{
+							data[i] = originData[j];
+							j++;
+						}
+					}
+				}
+
+				Size = size;
+				FrameSize = new Size(imageWidth, imageHeight);
+				this.offset = new float2(imageLeft, imageTop);
+
+				Data = data;
+			}
+		}
+
+		bool IsGif(Stream s)
+		{
+			var start = s.Position;
+
+			var a = s.ReadASCII(6);
+
+			s.Position = start;
+			return a.StartsWith("GIF");
+		}
+
+		GifFrame[] ParseFrames(Stream s)
+		{
+			var start = s.Position;
+			s.Position += 3;
+			var version = s.ReadASCII(3);
+			if (version != "87a" && version != "89a")
+				throw new Exception();
+
+			var width = s.ReadUInt16();
+			var height = s.ReadUInt16();
+			var size = new Size(width, height);
+
+			var GCT = s.ReadUInt8();
+			var BCIndex = s.ReadUInt8();
+			var PAR = s.ReadUInt8();
+
+			if (getBitValue(GCT, 0) != 0)
+			{
+				var GCEntries = 1 << ((GCT & 0x7) + 1);
+
+				// Skip over palette.
+				parseColorTable(s, GCEntries);
+			}
+
+			var frames = new List<GifFrame>();
+
+			while (true)
+			{
+				var block = s.ReadUInt8();
+
+				switch ((BlockTypes)block)
+				{
+					case BlockTypes.EXTENSION_BLOCK:
+						var type = (BlockTypes)s.ReadUInt8();
+
+						// We dont care about any of these, so we skip them.
+						switch (type)
+						{
+							case BlockTypes.GRAPHIC_EXTENSION_BLOCK :
+								var lengthG = s.ReadUInt8();
+								if (lengthG == 4)
+									s.Position += 5;
+								else
+									s.Position -= 2;
+								break;
+							case BlockTypes.COMMENT_EXTENSION_BLOCK :
+								readSubblock(s);
+								break;
+							case BlockTypes.PLAIN_TEXT_EXTENSION :
+								var lengthP = s.ReadUInt8();
+								s.Position += lengthP;
+								readSubblock(s);
+								break;
+							case BlockTypes.APPLICATION_EXTENSION_BLOCK :
+								var lengthA = s.ReadUInt8();
+								if (lengthA != 11)
+									throw new Exception("Bad app extension");
+
+								var identifier = s.ReadASCII(8);
+								s.Position += 3;
+								if (identifier == "NETSCAPE")
+									s.Position += 5;
+								else
+									readSubblock(s);
+								break;
+							default :
+								readSubblock(s);
+								break;
+						}
+						break;
+					case BlockTypes.IMAGE_DESCRIPTOR:
+						frames.Add(new GifFrame(s, size));
+						break;
+					case BlockTypes.TRAILER :
+						return frames.ToArray();
+					default :
+						Console.WriteLine("Bad data");
+						break;
+				}
+
+				// Shouldnt happen but if Trailer is missing at EOF we should bail.
+				if (s.Position >= s.Length)
+					break;
+			}
+
+			return frames.ToArray();
+		}
+
+		public bool TryParseSprite(Stream s, out ISpriteFrame[] frames)
+		{
+			if (!IsGif(s))
+			{
+				frames = null;
```

Comment:
Because its cheaper, also its the convention, its like this in all the other parsers.